### PR TITLE
Refactor cloud-init YAML generation to use YAML.dump

### PR DESCRIPTION
**Pull request by fdr**
Replace heredoc string interpolation with YAML.dump data structure serialization for write_user_data, meta-data, and network-config. Add comprehensive edge-case tests (YAML-special usernames, multi-line init scripts, empty public keys, multi-NIC, swap on/off).

### CODE REVIEW

**Comment by fdr** [comment]
Claude's role in this:

This was a mostly unsupervised loop via https://github.com/fdr/walk. It had @jeremyevans 's sketch as a seed, though I editorialized it (I find a `write` method that auto-serializes as yaml a bit too slick, but a separate method name working the same is okay). Claude's main ungratifying job was running it through integration tests with a live counter-party hetzner host using restricted API keys to do integration testing, and used that to capture the output into the tests.

This was eighteen permutations of kinds of input  (way more than a human could reasonably justify) with a full re-run even upon making slight adjustments, since, it's basically a background job, requiring little effort from me.

File: rhizome/host/lib/vm_setup.rb [hunk 1]
```diff
@@ -9,6 +9,7 @@
 require "openssl"
 require "base64"
 require "uri"
+require "shellwords"
```
Comment (jeremyevans) [inline rhizome/host/lib/vm_setup.rb:12]: What's the reason for requiring this and using `shelljoin`, as opposed to passing an array of arguments (which is what was done by the previous code)? There doesn't seem to be an advantage to this approach.

File: rhizome/host/lib/vm_path.rb [hunk 2]
```diff
@@ -90,5 +91,15 @@ def home(n)
...
     end
+
+    # Method serializing data to YAML and writing, e.g. #write_yaml_user_data
+    yaml_write_method_name = "write_yaml_" + method_name
+    fail "BUG" if method_defined?(yaml_write_method_name)
+
+    define_method yaml_write_method_name do |data, prefix: nil|
+      s = YAML.dump(data, line_width: -1)
+      s.sub!(/\A---/, prefix) if prefix
+      write(home(file_name), s)
```
Comment (jeremyevans) [inline rhizome/host/lib/vm_path.rb]: Conceptually DRYer, even if 1 additional character:

```suggestion
      send(write_method_name, s)
```

File: rhizome/host/lib/vm_setup.rb (hunk 1 above)
Comment (fdr) [reply rhizome/host/lib/vm_setup.rb:12]: Oh man, the way that it renders in yaml is both what you'd expect and insane, like

```
- nft
- add
- {
- table
[..continue for a whole page, one entry per list entry]
```

You could make it the list form [a, b, c] of yaml, that's an option I could use, but...that's why. It hits you in the face when reading the tests with materialized strings

**Comment by fdr** [comment]
> Since this also only handles the YAML files that used `yq` and not the other YAML files, I don't see an advantage over the patch I prepared for this. Were there any issues found with my patch?
> 
> The new tests look good. This rhizome code was definitely undertested before.

No, I just had it slimmed down, but I don't yet have a strong opinion on how much to batch into the roll-out (which has a material cost these days). You want to get more in in one shot? that seems like it could be fine...

File: rhizome/host/lib/vm_setup.rb (hunk 1 above)
Comment (jeremyevans) [reply rhizome/host/lib/vm_setup.rb:12]: `YAML.dump` in Ruby doesn't appear to support dumping lists with `[]`. 

I think that for testing, the long form (one line per argument) is actually preferable. I'd allowed me to easily find a bug in my initial implementation (unremoved `,`), which would have been more difficult if all of the options were in one line.

Forcing `shelljoin` on creation and shell parsing when actually running seems like unnecessary work on both ends. I don't think it causing any significant performance issues, though.

One possibility is using `JSON.pretty_generate` instead of `YAML.dump` (all valid JSON is valid YAML 1.2). That supports more customization options.

**Comment by jeremyevans** [comment]
> > Since this also only handles the YAML files that used `yq` and not the other YAML files, I don't see an advantage over the patch I prepared for this. Were there any issues found with my patch?
> > The new tests look good. This rhizome code was definitely undertested before.
> 
> No, I just had it slimmed down, but I don't yet have a strong opinion on how much to batch into the roll-out (which has a material cost these days). You want to get more in in one shot? that seems like it could be fine...

If you are asking whether we should also handle the non-`yq` YAML files at the same time, I don't see a current need for it, since we have no indications that there are issues in that code. That's why my patch didn't change those files.

File: rhizome/host/lib/vm_setup.rb (hunk 1 above)
Comment (fdr) [reply rhizome/host/lib/vm_setup.rb:12]: Oh man, no way, it's pages of stuff per single nft command! a very ungratifying problem to debug if you wanted to debug what NFT was doing (wrong)

Comment (fdr) [reply rhizome/host/lib/vm_setup.rb:12]: we don't need shelljoin though, just a string. that was an oops.

Comment (fdr) [reply rhizome/host/lib/vm_setup.rb:12]: oh wait no, we do, because the charactres in the original are complex. that's more a caller thing to get the literal string mode.

Comment (fdr) [reply rhizome/host/lib/vm_setup.rb:12]: here's what someone trying to figure out why their nft expression is a tiny bit off has to look forward to:

```
     bootcmd:                                                                                                                                                                           
     - - nft                                                                                                                                                                            
       - add                                                                                                                                                                            
       - table                                                                                                                                                                          
       - ip6                                                                                                                                                                            
       - filter                                                                                                                                                                         
     - - nft                                                                                                                                                                            
       - add                                                                                                                                                                            
       - chain                                                                                                                                                                          
       - ip6                                                                                                                                                                            
       - filter                                                                                                                                                                         
       - output                                                                                                                                                                         
       - "{"                                                                                                                                                                            
       - type                                                                                                                                                                           
       - filter                                                                                                                                                                         
       - hook                                                                                                                                                                           
       - output                                                                                                                                                                         
       - priority                                                                                                                                                                       
       - '0'                                                                                                                                                                            
       - ";"                                                                                                                                                                            
       - "}"                                                                                                                                                                            
     - - nft                                                                                                                                                                            
       - add                                                                                                                                                                            
       - rule                                                                                                                                                                           
       - ip6                                                                                                                                                                            
       - filter                                                                                                                                                                         
       - output                                                                                                                                                                         
       - ip6                                                                                                                                                                            
       - daddr                                                                                                                                                                          
       - fd00:0b1c:100d:5AFE::/64                                                                                                                                                       
       - meta                                                                                                                                                                           
       - skuid                                                                                                                                                                          
       - "!="                                                                                                                                                                           
       - '0'                                                                                                                                                                            
       - tcp                                                                                                                                                                            
       - flags                                                                                                                                                                          
       - syn                                                                                                                                                                            
       - reject                                                                                                                                                                         
       - with                                                                                                                                                                           
       - tcp                                                                                                                                                                            
       - reset 
   ```

Comment (fdr) [reply rhizome/host/lib/vm_setup.rb:12]: experimenting with json mode now with a custom formatter to get [a, few, elements]

Comment (fdr) [reply rhizome/host/lib/vm_setup.rb:12]: eh, I think we lose too much. it makes that better but makes other legibility (e.g. multiline strings) worse.

**Comment by fdr** [comment]
why does my job keep getting cancelled?.. blah

**Comment by fdr** [comment]
this got through vm stuff but died into postgres stuff, so I think e2e is just busted right now/flaky due to timeout, taking well over an hour.
